crypto/BRCryptoKey.c

```diff
@@ -0,0 +1,278 @@
+//
+//  BRCryptoKey.c
+//  BRCore
+//
+//  Created by Ed Gamble on 7/30/19.
+//  Copyright © 2019 Breadwinner AG. All rights reserved.
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+
+#include "BRCryptoKey.h"
+#include "BRKey.h"
+#include "BRBIP39Mnemonic.h"
+#include "BRBIP32Sequence.h"
+#include "BRKeyECIES.h"
+
+// We create an arbitary BRAddressParams - disconnected from any know BITCOIN, etc address params.
+// We do this until we know how the Key is used.  Right now, the only accessors are to get the
+// serialized form of the Key.  Is somebody expecting a particular encoding?  Is somebody expectring
+// to decode the serialization and then pass it to some other 'string-y' interface?  Until we know
+// what those other interfaces require, we don't know appropriate address parameters.
+//
+// Our serialization/deserialization works because we use the same address params.  If the above
+// questions are answered, the various cryptoKeyCreate*() functions will need to specify their
+// address params AND (AND AND) the serialization will need to WRITE OUT the specific params.
+//
+#define CRYPTO_ADDRESS_PARAMS  ((BRAddressParams) { \
+    3 + BITCOIN_PUBKEY_PREFIX,  \
+    3 + BITCOIN_SCRIPT_PREFIX,  \
+    3 + BITCOIN_PRIVKEY_PREFIX, \
+    "cry" \
```

Comment:
"cry"?

Reply:
This is an intentionally nonsensical (but workable) BRAddressParams definition.  For use when we don't have any other info.